### PR TITLE
[CSS-6503] Add OAuth support for non-charmed external clients

### PR DESCRIPTION
**Pull request by gtato**
This is the equivalent of the  PR opened on the [K8S charm](https://github.com/canonical/kafka-k8s-operator/pull/77), for adding the oauthbearer listener. 

In a few words this PR does the following:
- adds the oauth interface from hydra
- on a relation on that interface, it adds a sasl/oauthbearer listener
- it adds a unit test (using both a scram and oauthbearer client) to check if the config is generated correctly 
- the rest is about reorganizing the protocols and the mechanisms (cause it was a bit messy)
  - protocols are: sasl_plain, sasl_ssl and ssl
  - mechanisms are: scram-sha-512, oauthbearer, and ssl (this one makes no sense but it is for parity and it matches only with ssl as a protocol)

### CODE REVIEW

File: src/managers/config.py
Comment (marcoppenheimer) [inline src/managers/config.py:331]: **todo:** I don't love this string formatting, it's pretty obtuse with all the new-lines. I can see it tripping us up down the road. though I get where you were going with it being neater. 
IMO it's fine having a very long line in this case, as most editors wrap the text anyway.

Maybe do something like the following? Just to avoid all the new-line chars and multiple appends.

```python
long_oauth_properties = """
    prop.{var1}.{var2}='abc'
    other_prop.{var3}={var4}
"""

return [
    default_oauth_properties.format(var1='foo', var2='bar', var3='baz', var4='boo'),
]
```

File: src/managers/config.py
Comment (marcoppenheimer) [inline src/managers/config.py:358]: **praise:** Really like the strong type-hinting all over this PR!

**Comment by marcoppenheimer** [comment]
Overall, bar a relatively minor style choice, it looks good!
I'll wait for @zmraul to review though, as he was the one that originally did the feature, so will be able to provide a better review.

**Comment by deusebio** [comment]
@gtato @marcoppenheimer @zmraul 

Is this now ready for review? I see there are a number of conflicts. Also, I just want to make sure that this is the final implementation consistent with the spec [DA077](https://docs.google.com/document/d/1vJsMMe0k10an6m2WfLE-3jCeLDmqr4N3H1HsK5zGbh0/edit?usp=drive_link) before diving in (I see latest commit are preceding the hackathon...was there any code-wise lesson learned there?)

I'm happy to move this forward but I'd like to first understand where we stand

**Comment by gtato** [comment]
> @gtato @marcoppenheimer @zmraul
> 
> Is this now ready for review? I see there are a number of conflicts. Also, I just want to make sure that this is the final implementation consistent with the spec [DA077](https://docs.google.com/document/d/1vJsMMe0k10an6m2WfLE-3jCeLDmqr4N3H1HsK5zGbh0/edit?usp=drive_link) before diving in (I see latest commit are preceding the hackathon...was there any code-wise lesson learned there?)
> 
> I'm happy to move this forward but I'd like to first understand where we stand

I will update the PRs on both repos (this and the k8s) shortly.

File: src/managers/config.py
Comment (gtato) [inline src/managers/config.py:317]: i decided to split this pr from the one about the truststore since it might introduce a scope creep. For the moment i added 
`oauth.ssl.endpoint.identification.algorithm=""` which disables the certificate hostname verification, and should work for both google and hydra

File: src/core/models.py
Comment (deusebio) [inline src/core/models.py:469]: **suggestion** maybe one could use the data class already provided by the charm lib and its parsing [here](https://github.com/canonical/kafka-operator/pull/168/files#diff-747d8e831b4929f2bce44463876e22478664105f47ed7b3f05ca14182b26a250R320)

File: src/literals.py
Comment (deusebio) [inline src/literals.py:89]: **niptick** I would find a signature `dict[(AuthProtocol, AuthMechanism), Ports]` more appropriate, but it is not a strong point

File: src/managers/config.py
Comment (deusebio) [inline src/managers/config.py:191]: **question** there is no way to get this from `ops`? It seems weird that we need to access to some environment variable. I would have thought that ops would have provided some apis/bindings

File: src/managers/config.py
Comment (deusebio) [inline src/managers/config.py:301]: **suggestion** To avoid nesting and too many if statements, a more readable approach could be to first check whether we have a `oauth` relation, and if so, just rely on the assumption that a OAuth client exists, something like:

```
if not self.state.oauth_relation:
    return []

listener = [listener for listener in self.client_listeners if listener.mechanism.startswith("OAUTH")][0]

validation_cfg = ...

return [oauth_cfg]
```

File: src/managers/config.py
Comment (deusebio) [inline src/managers/config.py:324]: **todo** This may be one of the most important configuration, and I believe this may need to be configured somehow. I would possibly just factor this out into a constant, attribute or something to provide a bit more exposure to this. Right now it is a bit buried in this configuration list

**Review by deusebio — COMMENTED** [review]
Thanks @gtato for this! The code looks good! I have mostly nipticks and suggestions, nothing really major, code-wise. 

The only major comment would be to have (either or both):
1. Integration tests
2. Some documentation on how this can be setup in the form of an how-to guide, either in discourse or as PR description. 

I understand 1 can be tricky, and I'm happy to call it nice-to-have. But we should at the very least have some documentation on how this can be setup, such that it will then be on the product team to convert that into an integration tests (hopefully)

File: src/managers/config.py
Comment (gtato) [reply src/managers/config.py:191]: I asked this question at some point. I was told that's it at the moment, and that there is this issue opened:
https://github.com/canonical/operator/issues/1113

Comment (deusebio) [reply src/managers/config.py:191]: All right! You are definitely ahead of me on this :D 

Thanks for the link! Hope they had at least a discussion in Madrid. The "messiness" mentioned in the issue worries me a bit that we may need to give some support/attention to the part above. But this is out of this PR scope. Happy to resolve this!

**Comment by gtato** [comment]
> Thanks @gtato for this! The code looks good! I have mostly nipticks and suggestions, nothing really major, code-wise.
> 
> The only major comment would be to have (either or both):
> 
> 1. Integration tests
> 2. Some documentation on how this can be setup in the form of an how-to guide, either in discourse or as PR description.
> 
> I understand 1 can be tricky, and I'm happy to call it nice-to-have. But we should at the very least have some documentation on how this can be setup, such that it will then be on the product team to convert that into an integration tests (hopefully)

1. I really wanted to add an integration test, but it does require some time (to mock the idp) that I didn't manage to allocate in the previous cycle. In this one it is even harder as the event-bus epics are not even in the roadmap (apart some maintenance).

2. Makes sense. I think we (@zmraul and I) can make a discourse post with the two ways of setting this: with the oauth integrator and hydra.

**Comment by deusebio** [comment]
Yes, don't worry. As said, I'm happy to have a ticket for the product team to do this, as long as we have documentation.

File: src/core/models.py
Comment (marcoppenheimer) [reply src/core/models.py:469]: TBF I like that it's in line with the rest of the models here.

File: src/managers/config.py
Comment (marcoppenheimer) [reply src/managers/config.py:324]: Yeah, I wonder if an 'easy' breaking up of this might be to take out the entire definition of `oauth_cfg` and put that in a function, that takes `oauth_listener` as an arg, so we only call `self.generate_oauth_listener_config(listener=oauth_listener)` or something.

**Review by marcoppenheimer — APPROVED** [review]
Overall, I'm REALLY happy with this, it feels light-touch and you've done a fantastic job at integrating with the pre-existing style and patterns. We appreciate that a lot, especially for a first major contribution!

I left a few comments, but honestly I don't think they're blocking, more just style suggestions.

**Review by zmraul — CHANGES_REQUESTED** [review]
Very good! Left some minor comments.

The changes on `managers/config.py` are mostly design around how to declare the listeners and whatnot, so I'm happy with them.

File: src/core/models.py
Comment (gtato) [reply src/core/models.py:469]: @zmraul I did that at some point, once alone and once with Marc, but the `RelationState` is somehow tied to `Data` that comes from `data_platfrom_libs`, and the code was becoming more and more complicated for no good reason.

File: src/core/models.py
Comment (gtato) [reply src/core/models.py:124]: thanks for catching this... i did some refactoring and left this behind

File: src/core/models.py
Comment (marcoppenheimer) [reply src/core/models.py:469]: Yeah it was weird. Having a duck-typed model is OK I reckon, we may need more in the future when integrating other non-DP charm libs.

Comment (zmraul) [reply src/core/models.py:469]: got it, thanks

File: src/managers/config.py
Comment (gtato) [reply src/managers/config.py]: closing this since the formatting was updated and there are multiple comments on this sections

File: src/literals.py
Comment (gtato) [reply src/literals.py]: updated (i kinda like this better too)

File: src/managers/config.py
Comment (gtato) [reply src/managers/config.py:301]: updated as suggested

File: src/managers/config.py
Comment (gtato) [reply src/managers/config.py]: I extracted the claim, and fallback claim into two variables, but did not create a separate method. After the suggestion above to simplify this method, I thought that wouldn't be necessary anymore.

File: src/core/models.py
Comment (gtato) [reply src/core/models.py]: removed

File: metadata.yaml
Comment (gtato) [reply metadata.yaml]: done

File: src/core/models.py
Comment (gtato) [reply src/core/models.py:460]: I am letting this like this for now, due to split decision :smile:

Comment (deusebio) [reply src/core/models.py:460]: It was a suggestion, which is not meant to be blocking. Resolving now (and feel free to resolve suggestion that stays as such, hence not blocking)

**Comment by deusebio** [comment]
> Makes sense. I think we (@zmraul and I) can make a discourse post with the two ways of setting this: with the oauth integrator and hydra.

I'm happy with the changes. There are only a couple of points we need to resolve before approval:

* I checked the charmed kafka documentation [here](https://discourse.charmhub.io/t/charmed-kafka-documentation/10288) and there is no post around the integration. I seem to remember that maybe either you or raul had written something. Can you guys make sure that the docs are easily accessible? 
* Are the changes from https://github.com/canonical/kafka-operator/pull/197 included here? can we close the other PR as superseded by this one?

Once these are resolved, I'm happy to approve

**Comment by zmraul** [comment]
>     * Are the changes from [Oauth #197](https://github.com/canonical/kafka-operator/pull/197) included here? can we close the other PR as superseded by this one?

I believe they are. #197 was made for reference, so I can close it now.

**Review by zmraul — APPROVED** [review]

**Review by deusebio — APPROVED** [review]

**Comment by deusebio** [comment]
Great job @gtato !! The code looks very great! Thanks for this very valuable contribution! Once the tests pass, I'm happy to merge this!

**Comment by deusebio** [comment]
@zmraul I suppose we can close the other PR. would you do that?

**Comment by deusebio** [comment]
@gtato we have just landed a small PR to avoid the failing tests of `sync_docs`. Just rebase and I believe all tests should be green! Thanks!

**Pull request merged**
